### ISSUE TITLE
Can not get task output in pipeline

### ISSUE THREAD

**Issue by gsmlg** [tektoncd/pipeline#2617]
# Expected Behavior
Get task results

# Actual Behavior
unable to find result referenced by param "lint" in "report": Could not find result with name output for task run lint

# Steps to Reproduce the Problem

1. Run pipeline
task yaml
```yaml
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: yarn-run
  namespace: cicd
spec:
  params:
    - name: cmd
      type: string
      default: lint
  results:
    - name: output
      description: task run output
    - name: result
      description: task run result
  steps:
    - name: run
      workingDir: $(workspaces.repo.path)
      image: node:12-alpine
      script: |
        #!/bin/sh
        REPO=$(workspaces.repo.path)
        yarn run $(params.cmd) | tee $(results.output.path)
        echo $? | tee $(results.result.path)
      volumeMounts:
        - name: yarnrc
          mountPath: /tekton/home/.yarnrc
          subPath: config
        - name: yarn-cache
          mountPath: /yarn-cache
  volumes:
    - name: yarnrc
      configMap:
        name: yarnrc
    - name: yarn-cache
      persistentVolumeClaim:
        claimName: yarn-cache
  workspaces:
    - name: repo
```
2. Can not get results of lint output when output is very long
pipeline describe:
```
Name:              web-push-pipeline-run-8h7kg
Namespace:         cicd
Pipeline Ref:      zddi-web-ci-pipeline
Service Account:   pipeline-run-service
Timeout:           1h0m0s
Labels:
 tekton.dev/pipeline=zddi-web-ci-pipeline
 triggers.tekton.dev/eventlistener=gitlab-listener-interceptor
 triggers.tekton.dev/trigger=push-pipeline
 triggers.tekton.dev/triggers-eventid=xzlb8

🌡️  Status

STARTED          DURATION     STATUS
57 minutes ago   23 minutes   Failed(PipelineValidationFailed)

💌 Message

unable to find result referenced by param "lint" in "report": Could not find result with name output for task run lint

📦 Resources

 No resources

⚓ Params

 NAME               VALUE
 ∙ project_name     -------
 ∙ repository_url   ------
 ∙ revision         93cfe7b3657ebbaf7d147afe5eeba0cc4b636f2f
 ∙ ref              -------
 ∙ user_name       ------
 ∙ user_username   ------
 ∙ user_email    ------

🗂  Taskruns

 NAME                                          TASK NAME   STARTED          DURATION     STATUS
 ∙ web-push-pipeline-run-8h7kg-build-rf2sl     build       52 minutes ago   17 minutes   Succeeded
 ∙ web-push-pipeline-run-8h7kg-lint-cjwtm      lint        52 minutes ago   17 minutes   Succeeded
 ∙ web-push-pipeline-run-8h7kg-install-k482g   install     57 minutes ago   5 minutes    Succeeded
 ∙ web-push-pipeline-run-8h7kg-clone-cjhxs     clone       57 minutes ago   25 seconds   Succeeded
```

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T23:35:15Z", GoVersion:"go1.14.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.5+k3s1", GitCommit:"58ebdb2a2ec5318ca40649eb7bd31679cb679f71", GitTreeState:"clean", BuildDate:"2020-05-06T23:43:23Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

- Tekton Pipeline version: 
```
v0.12.0
```

**Comment by sbwsg** [comment]
Could you provide the `kubectl get -o yaml taskrun <name of taskrun>`?  I expect that the data for the result was simply too large.

I recommend using a [workspace](https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md#specifying-workspaces) for data that you know can be very large. Results have a limited size, which we document here:  https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md#storing-execution-results

> Note: The maximum size of a Task's results is limited by the container termination log feature of Kubernetes, as results are passed back to the controller via this mechanism. At present, the limit is "2048 bytes or 80 lines, whichever is smaller.". Results are written to the termination log encoded as JSON objects and Tekton uses those objects to pass additional information to the controller. As such, Task results are best suited for holding small amounts of data, such as commit SHAs, branch names, ephemeral space names, and so on.

There is also another limit in play: the size of objects stored in etcd.  TaskRuns store results inside their Status field. For very large pieces of data this will increase the size of the stored TaskRun to a point that is too big to store in etcd.  [I believe this limit is 1MB](https://stackoverflow.com/questions/53012798/kubernetes-configmap-size-limitation).

There might one day be changes in the way task results are implemented that lift this restriction but noone's looking to implement that right now as far as I'm aware.

**Comment by sbwsg** [comment]
I wonder if we should perhaps enforce a size limit via the entrypoint - if a Task tries to emit a result file that is larger than, say, 500KB, we should abandon it and exit with an error.

**Comment by vincent-pli** [comment]
Based on my test, the max size is 4096 , if more than that the string will be trim from beginning then destroy the `json` format, `json.Unmarshal` failed, then `result` lost.

Agree with @sbwsg , need validate.

**Comment by gsmlg** [comment]
My task output is too large that overflow the output data struct.
And this break out the result `json` format.
🙏

**Comment by vdemeester** [comment]
Seems like this is fixed by #2626, closing
